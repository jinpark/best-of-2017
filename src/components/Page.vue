<template>
  <swiper-slide>
    <div class="overlay"></div>
    <video class="bg-video" :src="videoUrl" :poster="posterUrl" preload muted loop></video>
    <div class="video-info">
      <h1 class="title">{{anime.title}}</h1>
      <h3 class="japanese-title">{{ anime.japaneseTitle }}</h3>
      <p class="summary"><b>Summary</b>: {{anime.summary}}</p>
      <p class="review"><b>Review</b>: {{anime.review}}</p>
      <button class="button is-text" @click="show"><icon name="youtube-play" scale="1.5"></icon></button>
      <button class="button is-text" @click="muteToggle"><icon name="volume-off" scale="1.5"></icon></button>
    </div>
  </swiper-slide>
</template>

<script>

export default {
  name: 'Page',
  props: ['anime'],
  computed: {
    video () {
      return this.$el.getElementsByTagName('video')[0]
    },
    videoUrl () {
      return `/static/${this.anime.filename}.mp4`
    },
    posterUrl () {
      return `/static/${this.anime.filename}.jpg`
    }
  },
  methods: {
    muteToggle () {
      if (this.video.muted) {
        this.video.muted = false
      } else {
        this.video.muted = true
      }
    },
    show () {
      this.video.muted = true
      this.$modal.show('youtube', {youtubeId: this.anime.youtubeId})
    }
  },
  data () {
    return {
    }
  }
}
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>

h1, h2 {
  font-weight: normal;
}

button {
  color: white !important;
}

video.bg-video {
  position: absolute;
  width: 100vw;
  height: 100vh;
  top: 0;
  left: 0;
  z-index: -2;
  object-fit: cover;
}

.overlay {
    display: block;
    height: 100%;
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    z-index: -1;
    background: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAQAAAAECAYAAACp8Z5+AAAAG0lEQVQYV2NkYGD4z8DAwMgABXAGNgGwSgwVAFbmAgXQdISfAAAAAElFTkSuQmCC);
}

.video-info {
  width: 85vw;
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  background: rgba(0, 0, 0, 0.5);
  padding: 10px;
  color: #ffffff;
}

.title {
  margin-bottom: 0.1em;
  color: #ffffff;
}

.japanese-title {
  margin: 0;
}

p.summary {
  padding-bottom: 0.5em;
}

p.review {
  padding-top: 0.5em;
  border-top: 1px solid white;
}
</style>
