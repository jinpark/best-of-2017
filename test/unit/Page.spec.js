import Page from '@/components/Page'
import { shallow } from 'vue-test-utils'
import { createRenderer } from 'vue-server-renderer'

describe('Page.vue', () => {
  let wrapper
  const anime = {
    "title": "A Silent Voice (Film)",
    "japaneseTitle": "聲の形",
    "filename": "asilentvoice",
    "summary": "A young man is ostracized by his classmates after he bullies a deaf girl to the point where she moves away. Years later, he sets off on a path for redemption.",
    "review": "Everyone has regrets and everyone deals with them differently. What's the right way to redeem yourself?",
    "youtubeId": "nfK6UgLra7g"
  }
  const $modal = {
    show: jest.fn()
  }

  beforeEach(() => {
    wrapper = shallow(Page, {propsData: {anime}, mocks: {$modal}})
  })

  it('is a swiper-slide', () => {
    expect(wrapper.is('swiper-slide')).toBe(true)
  })

  it('has correct background video data', () => {
    const video = wrapper.find('video')
    expect(video.hasAttribute('src', '/static/asilentvoice.mp4')).toBe(true)
    expect(video.hasAttribute('poster', '/static/asilentvoice.jpg')).toBe(true)
  })

  it('hides mute button by default', () => {
    const muteButton = wrapper.find('button.unmute')
    const unmuteButton = wrapper.find('button.mute')
    expect(muteButton.exists()).toBe(true)
    expect(unmuteButton.exists()).toBe(false)
  })

  it('toggles mute button when clicked', () => {
    expect(wrapper.find('button.unmute').exists()).toBe(true)
    expect(wrapper.find('button.mute').exists()).toBe(false)
    wrapper.find('button.unmute').trigger('click')
    expect(wrapper.find('button.unmute').exists()).toBe(false)
    expect(wrapper.find('button.mute').exists()).toBe(true)
  })

  it('starts modal when youtube button clicked and mutes all the things', () => {
    const youtubeButton = wrapper.find('button.youtube')
    youtubeButton.trigger('click')
    expect(wrapper.vm.isMuted).toBe(true)
    expect(wrapper.vm.video.muted).toBe(true)
    expect(wrapper.vm.$modal.show).toHaveBeenCalledWith('youtube', {youtubeId: anime.youtubeId})
  })

  it('matches snapshot', () => {
    const renderer = createRenderer()
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err)
      expect(str).toMatchSnapshot()
    })
  })
})
