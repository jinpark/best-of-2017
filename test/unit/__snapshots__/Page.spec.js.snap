// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page.vue matches snapshot 1`] = `
<swiper-slide>
    <div class="overlay"></div>
    <video src="/static/asilentvoice.mp4" poster="/static/asilentvoice.jpg" preload="" muted="muted" loop="loop" playsinline="" class="bg-video"></video>
    <div class="video-info">
        <h1 class="title">A Silent Voice (Film)</h1>
        <h3 class="japanese-title">聲の形</h3>
        <p class="summary">
            <b>Summary</b>: A young man is ostracized by his classmates after he bullies a deaf girl to the point where she moves away. Years later, he sets off on a path for redemption.</p>
        <div class="info">
            <p class="review">
                <b>Review</b>: Everyone has regrets and everyone deals with them differently. What's the right way to redeem yourself?</p>
            <button class="button is-text youtube">
                <icon name="youtube-play" scale="1.5"></icon>
            </button>
            <!---->
            <button class="button is-text unmute">
                <icon name="volume-off" scale="1.5"></icon>
            </button>
        </div>
    </div>
</swiper-slide>
`;
